stan/math/prim/fun/grad_pFq.hpp

```diff
@@ -0,0 +1,222 @@
+ *
+ * \f$ \frac{\partial}{\partial z}= \frac{\prod_{j=1}^pa_j}{\prod_{j=1}^qb_j}
+ *      {}_pF_q(a_1 + 1,...,a_p + 1; b_1 +1,...,b_q+1;z) \f$
+ *
+ * @tparam calc_a Boolean for whether to calculate derivatives wrt to 'a'
+ * @tparam calc_b Boolean for whether to calculate derivatives wrt to 'b'
+ * @tparam calc_z Boolean for whether to calculate derivatives wrt to 'z'
+ * @tparam TupleT Type of tuple containing objects to evaluate gradients into
+ * @tparam Ta Eigen type with either one row or column at compile time
+ * @tparam Tb Eigen type with either one row or column at compile time
+ * @tparam Tz Scalar type
+ * @param[in] grad_tuple Tuple of references to evaluate gradients into
+ * @param[in] a Vector of 'a' arguments to function
+ * @param[in] b Vector of 'b' arguments to function
+ * @param[in] z Scalar z argument
+ * @param[in] precision Convergence criteria for infinite sum
+ * @param[in] max_steps Maximum number of iterations for infinite sum
+ * @return Generalised hypergeometric function
+ */
+template <bool calc_a, bool calc_b, bool calc_z, typename TupleT, typename Ta,
+          typename Tb, typename Tz,
+          require_all_eigen_vector_t<Ta, Tb>* = nullptr,
+          require_stan_scalar_t<Tz>* = nullptr>
+void grad_pFq_impl(TupleT&& grad_tuple, const Ta& a, const Tb& b, const Tz& z,
+                   double precision, int max_steps) {
+  using std::max;
+  using scalar_t = return_type_t<Ta, Tb, Tz>;
+  using Ta_plain = plain_type_t<Ta>;
+  using Tb_plain = plain_type_t<Tb>;
+  using T_vec = Eigen::Matrix<scalar_t, -1, 1>;
+
+  Ta_plain ap1 = (a.array() + 1).matrix();
+  Tb_plain bp1 = (b.array() + 1).matrix();
+  Ta_plain log_a = log(a);
+  Tb_plain log_b = log(b);
+  scalar_type_t<Tz> log_z = log(z);
+  scalar_type_t<Ta> sum_log_a = sum(log_a);
+  scalar_type_t<Tb> sum_log_b = sum(log_b);
+
+  // Declare vectors to accumulate sums into
+  // where NEGATIVE_INFTY is zero on the log scale
+  T_vec da_infsum = T_vec::Constant(a.size(), NEGATIVE_INFTY);
+  T_vec db_infsum = T_vec::Constant(b.size(), NEGATIVE_INFTY);
+  // Vectors to accumulate outer sum into
+  T_vec da_iter_m = T_vec::Constant(a.size(), NEGATIVE_INFTY);
+  T_vec da_mn = T_vec::Constant(a.size(), NEGATIVE_INFTY);
+  T_vec db_iter_m = T_vec::Constant(b.size(), NEGATIVE_INFTY);
+  T_vec db_mn = T_vec::Constant(b.size(), NEGATIVE_INFTY);
+
+  // Only need the infinite sum for partials wrt p & b
+  if (calc_a || calc_b) {
+    double log_precision = log(precision);
+
+    int m = 0;
+    scalar_t outer_diff = 0;
+
+    da_infsum.setConstant(NEGATIVE_INFTY);
+    db_infsum.setConstant(NEGATIVE_INFTY);
+
+    while ((outer_diff > log_precision) && (m < max_steps)) {
+      // Vectors to accumulate outer sum into
+      da_iter_m.setConstant(NEGATIVE_INFTY);
+      da_mn.setConstant(NEGATIVE_INFTY);
+      db_iter_m.setConstant(NEGATIVE_INFTY);
+      db_mn.setConstant(NEGATIVE_INFTY);
+
+      double log_phammer_1m = log_rising_factorial(1, m);
```

Comment:
Since you're incrementing `m` by 1 at each step and I suggested not re-calculating this every iteration:

Rising factorial:
```
m = 0
rising_factorial = 1
while (m < max_steps) {
  term_m = (x + m);
  rising_factorial = rising_factorial * term_m;
  m++
}
```

log version: 
```
m = 0
log_rising_factorial = 0
while (m < max_steps) {
  term_m = log(x + m);
  log_rising_factorial = log_rising_factorial + term_m;
  m++
}
```

Something like that? You end up tracking the value and it's increment at each step.  Similarr situation with `lgamma`... if you go down this road I can be more helpful just ask and I'll write it as non-pseudocode

Reply:
This is a great idea! I didn't get it on first reading but now I understand, this should make things much more efficient. Thanks @sakrejda!